### PR TITLE
Use latest release instead of main branch on external tests

### PR DESCRIPTION
**Pull request by r0qs**
This PR modifies most of the external tests to utilize their latest releases, determined by the repository's Git tags, instead of their main branch. However, it retains the external tests that depend on specific branches or don't have releases/tags, such as `yield-liquidator-v2`.

### CODE REVIEW

File: scripts/externalTests/common.sh
Comment (cameel) [inline scripts/externalTests/common.sh]: But I'm assuming that it at least gets it right for the tests we have?

You could also make it look only at annotated tags. These are usually meant for releases because they have a date and description. Though I'm not sure if everyone sticks to that in practice.

File: test/externalTests/bleeps.sh
Comment (cameel) [inline test/externalTests/bleeps.sh]: Does it make any difference? The link without it looks cleaner so I'd leave it as is if it works.

File: scripts/externalTests/common.sh
Comment (cameel) [inline scripts/externalTests/common.sh]:
```suggestion
    [[ $ref != "" ]] || assertFail
```

Also, I see an unorthodox 3-char indent above. Are you trying to start a flame war? :P

File: scripts/externalTests/common.sh
Comment (cameel) [inline scripts/externalTests/common.sh:116]: Nice. I was going to comment about dropping `--depth 1`, potentially making the clone longer, but looks like this would make it a non-issue?

File: scripts/externalTests/common.sh
Comment (cameel) [reply scripts/externalTests/common.sh]: Also, you should assert that `$ref` is empty before overwriting it for `release`. If it's not empty better to fail than to ignore it silently.

File: test/externalTests/brink.sh
Comment (cameel) [inline test/externalTests/brink.sh]: The fact that `ref` is empty looks a bit confusing. Maybe it would be better to change `ref_type` to `latest_release`? Or require `ref=latest` or something like that?

File: scripts/externalTests/common.sh
Comment (r0qs) [reply scripts/externalTests/common.sh]: Yes, I was about to comment about it. I tried different release discovery mechanisms, mostly based on git tags or GitHub release/tags API, but all seem to produce wrong information of some sort, depending on how developers choose to create their tags/releases.

For instance, I tried to iterate over each tag and sort them by version numbers, returning the last one, but depending on the naming convention used by the repo, it can generate the wrong ordering.

Also, mechanisms that work in one repo may not work on another. For example, OpenZeppelin, even has a tag for an older release (`v4.9.5`) that is tagged after the latest release (`v5.0.1`), so a reverse list of tags sorted by chronological order returns the wrong latest release.

Using the REST GitHub API to retrieve the tag before cloning the repo may also not work for all cases either, for similar reasons.

I also created a script to check some variations of commands in an attempt to find one that fits all external repositories. The repos that have no output are because they have no tags or for option 6 because there is no GitHub release, so the curl failed with 404.

```
--- REPO: bleeps ---
TAGS:
1) bleeps_migrations
2) bleeps_migrations
3) sale
4) bleeps_migrations
5) sale
6)
--- REPO: brink ---
TAGS:
1) v1.1.0
2) v1.1.0
3) v1.1.0
4) v1.1.0
5) v1.1.0
6)
--- REPO: chainlink ---
TAGS:
1) v1.1.0
2) v1.1.0
3) v.0.8.12
4) v1.1.0
5) vtest-2020-03-03
6)
--- REPO: colony ---
TAGS:
1) burgundy-glider
2) burgundy-glider
3) rinkeby-v1
4) burgundy-glider
5) rinkeby-v1
6)
--- REPO: elementfi ---
TAGS:
1) v1.1.0
2) v1.1.0
3) v1.1.0
4) v1.1.0
5) v1.1.0
6) v1.1.0
--- REPO: ens ---
TAGS:
1) v1.1.5
2) v1.1.5
3) v.1.1.1-testnet
4) v1.1.5
5) v1.1.5-testnet
6) v1.1.5
--- REPO: euler ---
TAGS:
1) eip3-release
2) eip3-release
3) solidified-audit
4) eip3-release
5) solidified-audit
6)
--- REPO: gnosis ---
TAGS:
1) v1.4.1-build.0
2) v1.4.1-build.0
3) v1.4.1-build.0
4) v1.4.1-build.0
5) v1.4.1-build.0
6) v1.4.1-build.0
--- REPO: gp2 ---

TAGS:
1) v1.5.0
2) v1.5.0
3) v1.5.0
4) v0.0.1-alpha.11
5) v1.5.0
6)
--- REPO: perpetual-pools ---
TAGS:
1)
2)
3)
4)
5)
6)
--- REPO: pool-together ---
TAGS:
1) v1.3.1
2) v1.3.1
3) v1.3.1
4) v1.3.1
5) v1.3.1
6) v1.3.1
--- REPO: trident ---
TAGS:
1)
2)
3)
4)
5)
6)
--- REPO: uniswap ---
TAGS:
1) v1.0.0
2) v1.0.0
3) v1.0.0-rc.2
4) v1.0.0
5) v1.0.0-rc.2
6)
--- REPO: yield-liquidator ---
TAGS:
1)
2)
3)
4)
5)
6)
--- REPO: zeppelin ---
TAGS:
1) v4.8.0
2) v4.9.5
3) v5.0.1
4) v5.0.1
5) v5.0.1
6) v5.0.1
```

```bash
#!/bin/bash

SRC_DIR=$(pwd)
DIR=$(mktemp -d -t "ext-test-tags-check-XXXXXX")
cd $DIR
for repo_file in ${SRC_DIR}/test/externalTests/*.sh; do
        repo_url=$(sed -n "s|local repo=\"\(.*\)\"|\1|p" ${repo_file} | tr -d '[:space:]')
	filename=$(basename "${repo_file%.*}")
	echo "--- REPO: $filename --- "
	git clone --quiet --filter=blob:none $repo_url $filename 2> /dev/null
	(
	cd $filename
	tag1=$(git describe --tags --abbrev=0)
	tag2=$(git describe --tags $(git rev-list --tags --max-count=1))
	tag3=$(git for-each-ref \
			--format "%(refname:short)" \
			refs/tags \
			| sort --reverse --version-sort \
			| head --lines=1 \
	)
	tag4=$(git tag --sort=-authordate | head --lines=1)
	tag5=$(git tag --sort=-v:refname | head --lines=1)
	repo_location=$(echo "${repo_url}" | sed -E -n "s;.*github\.com[\/|:](.*).git;\1;p")
	tag6=$(curl -s -L --fail --show-error \
	 "https://api.github.com/repos/${repo_location}/releases/latest" | jq --raw-output .tag_name)
	printf "TAGS:\n1) ${tag1}\n2) ${tag2}\n3) ${tag3}\n4) ${tag4}\n5) ${tag5}\n6) ${tag6}\n"
	)
done
```

Comment (r0qs) [reply scripts/externalTests/common.sh]: > You could also make it look only at annotated tags. These are usually meant for releases because they have a date and description. Though I'm not sure if everyone sticks to that in practice.

Most of them don't do that unfortunately.

File: test/externalTests/bleeps.sh
Comment (r0qs) [reply test/externalTests/bleeps.sh]: No, I just wanted to follow the same standard that github uses. But I will remove it, no problem.

**Review by cameel — DISMISSED** [review]
Looks great!

I suggested some tweaks, but I'm going to pre-emptively approve it - i.e. I'm expecting that they'll be addressed, but they're not critical and the PR won't need another review pass. As soon as you think it's done and tests pass, it's ready to merge.

File: scripts/externalTests/common.sh
Comment (r0qs) [reply scripts/externalTests/common.sh]: hahaha, ops.

File: scripts/externalTests/common.sh
Comment (r0qs) [reply scripts/externalTests/common.sh:116]: With `--depth 1`, the commit history isn't retrieved, so it would have no tags. Using `blob:none` should be similarly fast to clone since it retrieves blobs on demand.

File: test/externalTests/brink.sh
Comment (r0qs) [reply test/externalTests/brink.sh]: Yeah, it could be `ref=latest` I guess. The idea is that it would actually never be empty, thus that assertion in the `common.sh` that you mentioned here https://github.com/ethereum/solidity/pull/14880#discussion_r1498052483. Or the `ref` was provided in the external test script or it was retrieved by the `git tag` command, otherwise something wrong happened that we should be notified.

File: scripts/externalTests/common.sh
Comment (cameel) [reply scripts/externalTests/common.sh]: Oh, it's a mess :) Looks like it was quite a bit of effort to even get that far. Seems like generically detecting the latest release is one of those fuzzy things that looks straightforward on the surface but then you run into all sorts of annoying corner cases.

Maybe sorting them via semver and rejecting those tags that do not validate would work better. It might also filter out prereleases or other weird tags. But that's another dependency so I think what you have is good enough too.

> `--- REPO: chainlink ---`
> `5) vtest-2020-03-03`

That one is unfortunate. If github's drop-down sorts tags chronologically then `v2.9.1-vrf-20240220` should be next and is actually up to date. Maybe an option to black-list some tags like `vtest-2020-03-03` would be useful? Alternatively, you could filter out all tags that do not start with `v<digit>`.

File: test/externalTests/brink.sh
Comment (cameel) [reply test/externalTests/brink.sh]: > The idea is that it would actually never be empty, thus that assertion in the `common.sh` that you mentioned here

Yeah, but you can still assert that it's `latest` and then overwrite it with the actual latest tag. You might also use something like `<latest>` to make it more apparent that it's just a placeholder.

Comment (r0qs) [reply test/externalTests/brink.sh]: Ok, I will do that. But I guess the assert need to be after the `ref` update anyway, because if the `git tag` for some reason returns empty, the `git checkout` with an empty string will be a no-op, and it should fail instead.

File: scripts/externalTests/common.sh
Comment (r0qs) [inline scripts/externalTests/common.sh]: Now that I'm thinking again about it, we could actually remove the `ref_type` completely.

File: scripts/externalTests/common.sh
Comment (r0qs) [reply scripts/externalTests/common.sh]: Honestly, not sure if it is worth the amount of work to workaround these stuffs.

Comment (ekpyron) [reply scripts/externalTests/common.sh]: The more complex the logic for this is, the more likely it'll break... if it's too complicated to reliably determine the latest release, we *can* also just hardcode it and use traditional methods to remind us to just regularly check/bump them... but if the current method works for the tests we have, I'd just stick with it for now.
But yeah, when suggesting to track the respective latest release automatically, I didn't think through the fact that that may in fact be quite complicated to do reliably in full generality - no need to go miles out of our way for it, I'd say. *Or* we could just define a "fetch latest release" mechanism in the respective script for the particular external test - which can then either use the simplest mechanism to fetch it (if any) for each external repo - or just a fixed one (if none). Really no need to try to solve the problem for every possible repo imaginable, is there :-)?

File: scripts/externalTests/common.sh
Comment (r0qs) [reply scripts/externalTests/common.sh]: I remove it in a separated commit, in case you think it is a good idea.

File: test/externalTests/brink.sh
Comment (cameel) [reply test/externalTests/brink.sh]: Well, you need both actually. Before update `ref` should be `<latest>` when `ref_type` is `release`. After the update you assert that it's non-empty.

I mean, unless you remove `ref_type`. Then you can't check that first condition any more. But then it does not matter because `<latest_release>` or something like that actually indicates that it's a release and no need to check anything else.

File: scripts/externalTests/common.sh
Comment (cameel) [inline scripts/externalTests/common.sh]:
```suggestion
    local repo_url="$1"
    local ref="$2"
    local test_dir="$3"
```

File: scripts/externalTests/common.sh
Comment (cameel) [inline scripts/externalTests/common.sh]: If you're removing `ref_type`, I'd make this more specific to make it clear it's the latest release not e.g. the latest commit:
```suggestion
    if [[ "$ref" == "<latest_release>" ]]; then
```

File: scripts/externalTests/common.sh
Comment (cameel) [reply scripts/externalTests/common.sh]: It's fine. The reason we had this parameter was that `git clone --branch <ref>` won't handle commits so I had to distinguish them and handle commits separately. If it's now all handled uniformly by the new implementation, we no longer have to know that so we can simplify things.

File: scripts/externalTests/common.sh
Comment (ekpyron) [reply scripts/externalTests/common.sh]: Actually, I'd myself personally even remove the ``download_project`` function entirely from ``common.sh`` and just do it manually in each of the external test script themselves, that looks to me like it'd be the much simpler choice and the amount of common code in this is quite negligible :-) - but I don't want to weigh in heavily on this from the side, you're the ones taking care of these scripts, so 100% your call :-).

**Review by cameel — APPROVED** [review]

File: scripts/externalTests/common.sh
Comment (r0qs) [reply scripts/externalTests/common.sh]: > The more complex the logic for this is, the more likely it'll break... if it's too complicated to reliably determine the latest release, we _can_ also just hardcode it and use traditional methods to remind us to just regularly check/bump them... but if the current method works for the tests we have, I'd just stick with it for now. But yeah, when suggesting to track the respective latest release automatically, I didn't think through the fact that that may in fact be quite complicated to do reliably in full generality - no need to go miles out of our way for it, I'd say. _Or_ we could just define a "fetch latest release" mechanism in the respective script for the particular external test - which can then either use the simplest mechanism to fetch it (if any) for each external repo - or just a fixed one (if none). Really no need to try to solve the problem for every possible repo imaginable, is there :-)?

I completely agree. I tried to find a command that could at least provide an intersection of the latest tags across our external tests only.


> Actually, I'd myself personally even remove the `download_project` function entirely from `common.sh` and just do it manually in each of the external test script themselves, that looks to me like it'd be the much simpler choice and the amount of common code in this is quite negligible :-) - but I don't want to weigh in heavily on this from the side, you're the ones taking care of these scripts, so 100% your call :-).

Yes, I mean, it's already done, so let's leave it as it is for now. We'll eventually migrate everything to Python anyway, and I believe the maintenance burden will be much less.

Comment (cameel) [reply scripts/externalTests/common.sh]: ok, if you think it adds too much complexity let's forget about blacklisting tags and keep Chainlink at `master`. If we did add it, I'd still prefer to hide that in `download_project` though - the idea was that most of the things in the project-specific scripts are just invocations of reusable functionality. You can mix and match any steps from the toolbox you need, while still keeping the recipe concise. Unfortunately with this being written in Bash, it sill ends up being quite messy, but since we'll eventually rewrite it in Python, all the more reason not to change the whole design right now.

Also, I just noticed that blacklisting tags for Chainlink would be pointless after all, because we use our fork from `solidity-external-tests` and not the upstream. So problem solved :)

**Pull request merged**
